src/jit/hwintrinsicxarch.cpp

```diff
@@ -295,6 +304,163 @@ GenTree* Compiler::getArgForHWIntrinsic(var_types argType, CORINFO_CLASS_HANDLE
     return arg;
 }
 
+//------------------------------------------------------------------------
+// immUpperBoundOfHWIntrinsic: get the max imm-value of non-full-range IMM intrinsic
+//
+// Arguments:
+//    intrinsic  -- intrinsic ID
+//    baseType   -- base type of the Vector128/256<T>
+//
+// Return Value:
+//     the max imm-value of non-full-range IMM intrinsic
+//
+int Compiler::immUpperBoundOfHWIntrinsic(NamedIntrinsic intrinsic, var_types baseType)
+{
+    assert(categoryOfHWIntrinsic(intrinsic) == HW_Category_IMM);
+    HWIntrinsicFlag flags = flagsOfHWIntrinsic(intrinsic);
+    if ((flags & HW_Flag_FullRangeIMM) != 0)
+    {
+        return 255;
+    }
+
+    switch (intrinsic)
+    {
+        case NI_AVX_Compare:
+        case NI_AVX_CompareScalar:
+            return 31; // enum FloatComparisonMode has 32 values
+
+        case NI_SSE2_ShiftLeftLogical:
+        case NI_SSE2_ShiftRightArithmetic:
+        case NI_SSE2_ShiftRightLogical:
+        case NI_AVX2_ShiftLeftLogical:
+        case NI_AVX2_ShiftRightArithmetic:
+        case NI_AVX2_ShiftRightLogical:
+            return genTypeSize(baseType) * 8 - 1;
+
+        case NI_SSE2_ShiftLeftLogical128BitLane:
+        case NI_SSE2_ShiftRightLogical128BitLane:
+        case NI_AVX2_ShiftLeftLogical128BitLane:
+        case NI_AVX2_ShiftRightLogical128BitLane:
+            return 15;
+
+        default:
+            return 255;
+    }
+}
+
+//------------------------------------------------------------------------
+// impOutOfRangeFallback: convert certain SSE2/AVX2 shift intrinsic to its semantic alternative when the imm-arg is
+// not in the valid range
+//
+// Arguments:
+//    intrinsic  -- intrinsic ID
+//    simdType   -- Vector type
+//    baseType   -- base type of the Vector128/256<T>
+//
+// Return Value:
+//     return the IR of semantic alternative on invalid imm-arg values
+//
+GenTree* Compiler::impOutOfRangeFallback(NamedIntrinsic intrinsic, var_types simdType, var_types baseType)
+{
+    assert((flagsOfHWIntrinsic(intrinsic) & HW_Flag_OutOfRangeFallbackIMM) != 0);
+    assert(simdType == TYP_SIMD32 || simdType == TYP_SIMD16);
```

Comment:
I see that the formatter doesn't flag these but the usual practice is to parenthesize this kind of expressions: `assert((simdType == TYP_SIMD32) || (simdType == TYP_SIMD16));`. Happens in multiple places.